x/merkledb/trieview.go

```diff
@@ -192,52 +186,47 @@ func newTrieViewWithChanges(
 }
 
 // Recalculates the node IDs for all changed nodes in the trie.
-// Assumes [t.lock] is held.
 func (t *trieView) calculateNodeIDs(ctx context.Context) error {
-	switch {
-	case t.isInvalid():
-		return ErrInvalid
-	case !t.needsRecalculation:
-		return nil
-	case t.committed:
-		// Note that this should never happen. If a view is committed, it should
-		// never be edited, so [t.needsRecalculation] should always be false.
-		return ErrCommitted
-	}
-
-	// We wait to create the span until after checking that we need to actually
-	// calculateNodeIDs to make traces more useful (otherwise there may be a span
-	// per key modified even though IDs are not re-calculated).
-	ctx, span := t.db.tracer.Start(ctx, "MerkleDB.trieview.calculateNodeIDs")
-	defer span.End()
-
-	// ensure that the view under this one is up-to-date before potentially pulling in nodes from it
-	// getting the Merkle root forces any unupdated nodes to recalculate their ids
-	if _, err := t.getParentTrie().GetMerkleRoot(ctx); err != nil {
-		return err
-	}
+	var err error
+	t.calculateNodesOnce.Do(func() {
+		switch {
+		case t.isInvalid():
+			err = ErrInvalid
+			return
+		case t.committed:
+			// Note that this should never happen. If a view is committed, it should
+			// have already had its nodes calculated
+			err = ErrCommitted
+			return
+		}
 
-	_, helperSpan := t.db.tracer.Start(ctx, "MerkleDB.trieview.calculateNodeIDsHelper")
-	defer helperSpan.End()
+		// We wait to create the span until after checking that we need to actually
+		// calculateNodeIDs to make traces more useful (otherwise there may be a span
+		// per key modified even though IDs are not re-calculated).
+		ctx, span := t.db.tracer.Start(ctx, "MerkleDB.trieview.calculateNodeIDs")
+		defer span.End()
 
-	// [eg] limits the number of goroutines we start.
-	var eg errgroup.Group
-	eg.SetLimit(numCPU)
-	if err := t.calculateNodeIDsHelper(ctx, t.root, &eg); err != nil {
-		return err
-	}
-	if err := eg.Wait(); err != nil {
-		return err
-	}
-	t.needsRecalculation = false
-	t.changes.rootID = t.root.id
+		_, helperSpan := t.db.tracer.Start(ctx, "MerkleDB.trieview.calculateNodeIDsHelper")
+		defer helperSpan.End()
```

Comment:
I don't think we need this span anymore since it does the same thing as the span above